rir/src/ir/Compiler.cpp

```diff
@@ -376,40 +376,56 @@ bool compileSpecialCall(Context& ctx, SEXP ast, SEXP fun, SEXP args_) {
         // 3) Special case [ and [[
         if (lhsParts.size() == 2) {
             RList g(lhsParts[0]);
-            if (g.length() == 3) {
+            if (g.length() == 3 || g.length() == 4) {
+                bool is2d = g.length() == 4;
                 SEXP fun2 = *g.begin();
                 auto idx = g.begin() + 2;
-                if (*idx != R_DotsSymbol && *idx != R_MissingArg &&
-                    !idx.hasTag()) {
-                    if (fun2 == symbol::DoubleBracket ||
-                        fun2 == symbol::Bracket) {
-
-                        cs << BC::guardNamePrimitive(fun);
-
-                        // First rhs (assign is right-associative)
-                        compileExpr(ctx, rhs);
-                        // Keep a copy of rhs since its the result of this
-                        // expression
-                        cs << BC::dup() << BC::ensureNamed();
-
-                        // Now load index and target
-                        cs << BC::ldvar(target);
-                        compileExpr(ctx, *idx);
-
-                        // do the thing
-                        cs << BC::recordBinop();
-                        if (fun2 == symbol::DoubleBracket)
-                            cs << BC::subassign2();
-                        else
-                            cs << BC::subassign1();
-                        cs.addSrc(ast);
-
-                        // store the result as "target"
-                        cs << BC::stvar(target);
-
-                        cs << BC::invisible();
-                        return true;
+                auto idx2 = is2d ? (g.begin() + 3) : idx;
+                if ((fun2 == symbol::DoubleBracket ||
+                     // TODO: Figure out why 2d bracket assign fails sometimes
+                     // (specifically, f19 in tests/pir-assign)
+                     (!is2d && fun2 == symbol::Bracket)) &&
+                    *idx != R_DotsSymbol && *idx != R_MissingArg &&
+                    !idx.hasTag() &&
+                    (!is2d || (*idx2 != R_DotsSymbol && *idx2 != R_MissingArg &&
+                               !idx2.hasTag()))) {
```

Comment:
This is very confusing.. but then again, so is the whole compiler

Reply:
Checks that the indices aren't `...`, missing, or labeled (e.g. `a[foo = bar,...]` would fail). Doesn't check `idx2` unless it's a 2d subassign, because otherwise `idx2` isn't used